apm-agent-core/src/test/java/co/elastic/apm/agent/impl/MetaDataTest.java

```diff
@@ -0,0 +1,149 @@
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import static co.elastic.apm.agent.configuration.CoreConfiguration.CloudProvider.AUTO;
+import static co.elastic.apm.agent.configuration.CoreConfiguration.CloudProvider.AWS;
+import static co.elastic.apm.agent.configuration.CoreConfiguration.CloudProvider.GCP;
+import static co.elastic.apm.agent.configuration.CoreConfiguration.CloudProvider.NONE;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.when;
+
+class MetaDataTest {
+
+    private static ConfigurationRegistry config;
+    private static CoreConfiguration coreConfiguration;
+    @Nullable
+    private static CoreConfiguration.CloudProvider currentCloudProvider;
+
+    @BeforeAll
+    static void setup() {
+        config = SpyConfiguration.createSpyConfig();
+        coreConfiguration = config.getConfig(CoreConfiguration.class);
+        CloudProviderInfo cloudProviderInfo = CloudMetadataProvider.fetchAndParseCloudProviderInfo(AUTO, 1000);
+        if (cloudProviderInfo != null) {
+            currentCloudProvider = CoreConfiguration.CloudProvider.valueOf(cloudProviderInfo.getProvider().toUpperCase());
+        }
+    }
+
+    @BeforeEach
+    void reset() {
+        SpyConfiguration.reset(config);
+    }
+
+    @Test
+    void testCloudProvider_NONE() throws InterruptedException, ExecutionException, TimeoutException {
+        // The default configuration for cloud_provide is NONE
+        Future<MetaData> metaDataFuture = MetaData.create(config, null);
+        assertThat(metaDataFuture).isInstanceOf(MetaData.NoWaitFuture.class);
+        MetaData metaData = metaDataFuture.get(0, TimeUnit.MILLISECONDS);
+        verifyMetaData(metaData, NONE);
+    }
+
+    @ParameterizedTest
+    @EnumSource(value = CoreConfiguration.CloudProvider.class, names = {"AWS", "GCP", "AZURE"})
+    void testCloudProvider_SingleProvider(CoreConfiguration.CloudProvider provider) throws InterruptedException, ExecutionException, TimeoutException {
+        when(coreConfiguration.getCloudProvider()).thenReturn(provider);
+        Future<MetaData> metaDataFuture = MetaData.create(config, null);
+        assertThat(metaDataFuture).isNotInstanceOf(MetaData.NoWaitFuture.class);
+        Exception timeoutException = null;
+        try {
+            metaDataFuture.get(0, TimeUnit.MILLISECONDS);
+        } catch (TimeoutException e) {
+            timeoutException = e;
+        }
+        if (provider != GCP && provider != currentCloudProvider) {
+            // The GCP fails quickly on DNS lookup
+            assertThat(timeoutException).isInstanceOf(TimeoutException.class);
+        }
+        // In AWS we may need two timeouts - one for the API token and one for the metadata itself
+        long timeout = (long) (coreConfiguration.geCloudMetadataDiscoveryTimeoutMs() * ((provider == AWS) ? 2.5 : 1.5));
```

Comment:
[question] would there be a way to avoid static timeouts as those might imply test flakyness ?

Reply:
I don't think this implies flakiness, it is a **timeout** with a big safety cushion - discovery of AWS is very quick on AWS and almost exactly 2 seconds elsewhere, with Azure it would be quick on Azure and about 1 second elsewhere. It was my test confirming that all metadata discovery queries occur in parallel. If we see it is flaky, it's extremely easy to just change the timeout to 5 seconds, but a test with a timeout is necessary as we want to make sure nobody breaks the metadata discovery timeouts and let them just execute for too long.